### DISCUSSION TITLE
Separate channels with macthing version numbers

### DISCUSSION THREAD

**Discussion by derrod** [sparkle-project/Sparkle#2283] [Q&A] [answered]
Hi all,

for OBS Studio we wanted to (finally) upgrade to Sparkle 2 and make use of the channels feature to allow users to opt-into receiving beta updates.

In our case we do not provide universal builds due to size concerns (CEF is yuge), so we instead have separate arm64/x86_64 builds that previously have used separate feeds. With channels in Sparkle 2 we hoped to provide a unified feed where we distinguish arm64/x86 with suffixes on their channel names (e.g. `stable_arm64`, `beta_amd64`, etc.).

In this process I ran into two issues/edge cases that I wanted to ask about/clarify since I'm not sure if they warrant opening an issue (yet):
1. `generate_appcast` does not allow two items to have the same version, even if they are on different channels
    * Rather than adding a new item, the existing item on channel "A" will be updated even when channel "B" is specified in the command line (though the channel attribute of the existing item won't be updated)
3. `generate_appcast` will generate deltas between channels, and there doesn't seem to be a way to disable this (or delta generation in general)?

So I wanted to clarify whether either of these would be considered a bug/feature request or working as intended. As far as I can tell there shouldn't be any issue with two items having the same version as long as they are on different channels.

Having the same version on two separate channels seems like something that should be allowed, not just for cases where you have separate architectures but also to deliver a stable build to a beta channel when no betas are currently available (we have our own implementation to tell the client which channels are currently available to handle this fallback).

Thanks

**Comment by zorgiepoo** [comment, marked as answer]
I’m typing this in a slight hurry but channels are not intended for parallel trains. In this case you’ll want to still use separate appcasts.

There is also separately the issue of not providing a universal binary as this is not the “Apple/macOS way” and it confuses users. A couple of things like delta updates and file system compression can help with download and disk space concerns a bit. 

> Having the same version on two separate channels seems like something that should be allowed, not just for cases where you have separate architectures but also to deliver a stable build to a beta channel when no betas are currently available 

A stable build would be delivered to the default channel, which any client subscribing to the beta channel would also see.

**Reply by derrod** [reply]
Alright, thank you for the swift reply!

Continuing to use separate appcasts is simple enough. We'll just have to build something to also generate 1.x compatible feeds so our old versions can update to the new stable ones.

**Reply by zorgiepoo** [reply]
[Upgrading to newer features](https://sparkle-project.org/documentation/publishing/#upgrading-to-newer-features) describes switching to a new appcast if you want to ensure application versions using this appcast have access to newer features (like channels).

**Reply by zorgiepoo** [reply]
Also the top of that page describes compression approaches that may sway you not splitting the archs.. Chrome itself (nowadays) does not split its architectures.

**Reply by zorgiepoo** [reply]
@derrod Unrelated, I was looking at the OBS PR very briefly. If you want to leave Sparkle configured such that Sparkle manages checking for updates in the background automatically, and first asks the user's permission if they want to check for updates automatically (this is the default configuration) then you probably do not want to manually call `checkForUpdatesInBackground` (this is how OBS currently operates). I am even considering adding a small warning about this in #2295

**Reply by derrod** [reply]
In the OBS PR `checkForUpdatesInBackground` is actually only used when a user switches the update channel they have selected. Otherwise it is left up to Sparkle to check for updates on startup, and manually triggering the check uses `checkForUpdates`.

That said the setup is probably still a bit awkward given that we also have our own checkbox for enabling checking for updates, and I shall revisit that.

**Reply by zorgiepoo** [reply]
Ah that is fine then.

> That said the setup is probably still a bit awkward given that we also have our own checkbox for enabling checking for updates, and I shall revisit that.

Apps would use or set `updater.automaticallyChecksForUpdates` for such a checkbox (edit: I forgot our documentation has [an example here](https://sparkle-project.org/documentation/preferences-ui#adding-settings-in-qt) which you may have already seen).
